### PR TITLE
feat(secrets): JIT Snyk Learn lessons for Secrets findings [EDU-4754]

### PR DESCRIPTION
**Pull request by SnykOleg**
## Why

Secrets findings in the IDE had no Snyk Learn JIT link on the **issue card** (unlike SAST/OSS). This adds CWE-798 lesson lookup during scan and product-aligned panel copy so IDE parity matches the web inventory work in [EDU-4753](https://snyksec.atlassian.net/browse/EDU-4753) / [app-ui#5425](https://github.com/snyk/app-ui/pull/5425).

Jira: [EDU-4754](https://snyksec.atlassian.net/browse/EDU-4754). **Learn code actions** (editor lightbulb) are **out of scope**.

---

## Product review (Waleed / Jakob)

**What you're reviewing:** the **Secrets issue details panel** in the IDE (HTML from snyk-ls when you open a finding from the tree). Not the web inventory drawer ([app-ui#5425](https://github.com/snyk/app-ui/pull/5425) / [#5426](https://github.com/snyk/app-ui/pull/5426)).

**What changed for users**

- Secrets issue details can show a **Snyk Learn** link when the finding maps to CWE-798 (Hardcoded secrets lesson).
- Copy: **"Learn how to remediate Secrets securely"** (secrets-specific; not the SAST string "Learn about this issue type").
- Header uses **Location:** with a clickable line number.
- Multiple locations in the **same file** show **`N LOCATIONS IN THIS FILE`**.
- Existing **Details and remediation** accordion stays (static guidance text).

**Previews:** LS HTML with simulated VS Code Dark+ theme. Final spacing/fonts depend on the IDE webview.

### Visuals

| Scenario | What to verify | Preview |
|----------|----------------|---------|
| Learn link present (CWE-798) | CTA text, external link icon, placement under header | panel-with-learn |
| Learn link absent (no lesson resolved) | No Learn block; rest of panel unchanged | *(no screenshot — panel shows **Location:** + Details accordion only; no Learn CTA between them)* |
| Multi-location, same file | Banner count matches locations in file | panel-multi-location |

### IDE vs web (EDU-4753)

| Element | IDE (this PR) | Web inventory |
|---------|---------------|---------------|
| Learn JIT | Issue card link | Drawer (#5426) |
| Learn copy | Learn how to remediate Secrets securely | Align with #5426 |
| Location | **Location:** + line link | Single Location row |
| Remediation body | Collapsible in card | How-to-Fix hidden (#5425) |
| Score / table sort | N/A in card | Web-only |

### Out of scope

- Learn **code actions** (editor lightbulb) for secrets
- IDE plugin changes (HTML panel only)
- Web drawer score banner, table sort, single Location row

### Product sign-off checklist

- [ ] Learn CTA copy approved
- [ ] **Location:** label matches web convention
- [ ] Multi-location banner acceptable when 2+ hits in same file
- [ ] No Learn link case acceptable (fail-soft, no error shown)

### Optional live check

1. Scan a repo with a CWE-798 secret (e.g. fake-leaks).
2. Open the finding in the IDE tree → confirm panel shows Location + Learn link when lesson resolves.

---

<details>
<summary>Engineering details (for snyk-ls reviewers)</summary>

## Summary

CWE-keyed lookup resolves the published "Hardcoded secrets" lesson for findings tagged with CWE-798 (first CWE only, same as SAST/OSS).

1. Enable `learn.Service.GetLesson` for `types.SecretsIssue` in `lessonsLookupParams`.
2. Populate `Issue.LessonUrl` in `FindingsConverter.findingToIssues` (single conversion pass).
3. Render **"Learn how to remediate Secrets securely"** in the secrets details panel.
4. Inject learn via `secrets.Scanner.enrichContext` so **autosave / background** scans get `LessonUrl`, not only `workspace/executeCommand` scans.
5. Document `SecretsIssue = 5` in README for `snyk.getLearnLesson` / `snyk.openLearnLesson`.
6. Product alignment: **Location:** label; per-file `LocationsCount` for the multi-location banner.

**Net diff vs `main`:** `infrastructure/learn/*`, `infrastructure/secrets/*`, `domain/ide/command/*_learn_lesson_test.go`, `application/di/init.go`, `README.md`, `docs/ui-rendering.md`.

**Fail-soft:** Learn lookup errors, nil lessons, or missing learn service leave `LessonUrl` empty. Scans never fail because of Learn.

```mermaid
flowchart LR
  scan[secrets Scan] --> enrich[enrichContext]
  enrich --> convert[FindingsConverter.findingToIssues]
  convert --> lesson[learn.GetLesson CWE-798]
  lesson --> panel[details.html LessonUrl]
```

### Companion work

- **Web (EDU-4753):** [app-ui#5426](https://github.com/snyk/app-ui/pull/5426) (stacked on [#5425](https://github.com/snyk/app-ui/pull/5425)). Parallel surface; not required to merge this ls PR.

### Notes for the reviewer

- **@bastiandoetsch feedback** addressed: Learn via request context; lesson lookup fused into `FindingsConverter.findingToIssues`.
- **SAST** still uses struct field + post-scan enhancer; aligning SAST to this pattern is a separate concern.

### Test plan

- [x] `make lint` — 0 issues
- [x] `make test` — pass (`infrastructure/learn`, `infrastructure/secrets`, `domain/ide/command`)
- [x] **Learn lookup:** `lessonsLookupParams` for `SecretsIssue` — `infrastructure/learn/service_test.go`
- [x] **Conversion / scan / HTML:** `convert_test.go`, `secrets_test.go`, `secrets_html_test.go`
- [x] **Commands:** `getLearnLesson` / `openLearnLesson` accept `issueType=5`
- [x] **Docs:** [`docs/ui-rendering.md`](docs/ui-rendering.md) (Secrets Issue Details Panel section)

### Backwards compatibility

`types.SecretsIssue` ordinal is `5` (already on main). `LS_PROTOCOL_VERSION` unchanged.

### Checklist

- [x] Tests added and all succeed
- [x] Regenerated mocks (`make generate`)
- [x] Linted (`make lint`)
- [x] README.md updated (`SecretsIssue = 5`)
- [x] Docs updated (`docs/ui-rendering.md`)

</details>

[EDU-4753]: https://snyksec.atlassian.net/browse/EDU-4753?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[EDU-4754]: https://snyksec.atlassian.net/browse/EDU-4754?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: infrastructure/secrets/secrets.go [hunk 1]
```diff
@@ -192,11 +195,41 @@ func (sc *Scanner) Scan(ctx context.Context, pathToScan types.FilePath) (issues
...
+// enhanceWithLearnLesson populates each secret issue's LessonUrl from the Snyk Learn
+// service. Mirrors infrastructure/code/code.go:enhanceIssuesDetails for the secrets product.
+// Errors are logged and swallowed; a missing or empty lesson leaves LessonUrl untouched.
+//
+// Sentry-reporter parity with SAST is intentionally omitted: secrets.Scanner has no
+// errorReporter field today and the rest of the package logs transient errors via the
+// scanner's zerolog logger only. A non-actionable Learn-cache miss does not warrant
+// expanding the constructor signature; revisit alongside any future Sentry pass on the
+// secrets package.
+func (sc *Scanner) enhanceWithLearnLesson(issues []types.Issue) {
```
Comment (SnykOleg) [inline infrastructure/secrets/secrets.go]: Heads-up for reviewers: the deliberate Sentry omission here is intentional, not an oversight. `secrets.Scanner` has no `errorReporter` field today and the rest of the package logs transient errors via the scanner's zerolog logger only. Adding `error_reporting.ErrorReporter` just for one non-actionable Learn-cache miss would expand the constructor signature disproportionately. Full rationale in the PR description under 'Implementation notes'. Revisit alongside any future Sentry pass on the secrets package.

File: infrastructure/secrets/secrets.go [hunk 2]
```diff
@@ -77,10 +79,11 @@ type Scanner struct {
 	configResolver     types.ConfigResolverInterface
 }
 
-func New(conf configuration.Configuration, engine workflow.Engine, logger *zerolog.Logger, instrumentor performance.Instrumentor, apiClient snyk_api.SnykApiClient, featureFlagService featureflag.Service, notifier notification.Notifier, configResolver types.ConfigResolverInterface) *Scanner {
+func New(conf configuration.Configuration, engine workflow.Engine, logger *zerolog.Logger, instrumentor performance.Instrumentor, apiClient snyk_api.SnykApiClient, learnService learn.Service, featureFlagService featureflag.Service, notifier notification.Notifier, configResolver types.ConfigResolverInterface) *Scanner {
 	return &Scanner{
 		IssueCache:         issuecache.NewIssueCache(product.ProductSecrets),
 		SnykApiClient:      apiClient,
+		learnService:       learnService,
```
Comment (bastiandoetsch) [inline infrastructure/secrets/secrets.go]: Instead of injecting a field here, please enrich the call context with learnService, ideally already when it's created. References: OSS scanner: https://github.com/snyk/snyk-ls/blob/270c5ad459dd3c2e823db97b07b97b675f4e914c/infrastructure/oss/cli_scanner.go#L708

Where it actually should be enriched: https://github.com/snyk/snyk-ls/blob/3e55892c053807177bef9e006ae93474d288efb9/application/server/server.go#L121

File: infrastructure/secrets/secrets.go [hunk 3]
```diff
@@ -192,11 +195,41 @@ func (sc *Scanner) Scan(ctx context.Context, pathToScan types.FilePath) (issues
...
+// Errors are logged and swallowed; a missing or empty lesson leaves LessonUrl untouched.
+//
+// Sentry-reporter parity with SAST is intentionally omitted: secrets.Scanner has no
+// errorReporter field today and the rest of the package logs transient errors via the
+// scanner's zerolog logger only. A non-actionable Learn-cache miss does not warrant
+// expanding the constructor signature; revisit alongside any future Sentry pass on the
+// secrets package.
+func (sc *Scanner) enhanceWithLearnLesson(issues []types.Issue) {
+	logger := sc.logger.With().Str("method", "secrets.enhanceWithLearnLesson").Logger()
+	for i := range issues {
```
Comment (bastiandoetsch) [inline infrastructure/secrets/secrets.go]: This is an additional loop over the issues. Can this be integrated into the conversion, please? We currently try to tackle performance issues, and having several loops over the same data is one of the patterns that is unfortunate.

File: infrastructure/secrets/secrets.go (hunk 2 above)
Comment (SnykOleg) [reply infrastructure/secrets/secrets.go]: Addressed in `9e8df31f` — `learn.Service` is now plumbed through `application/server/server.go:withContext` (mirroring the existing `authenticationService` / `ldxSyncService` / `inlineValueProvider` injection), and `secrets.Scanner` no longer carries a `learnService` field — the constructor reverts to its pre-PR 8-arg signature. The converter reads the service from `ctx2.DependenciesFromContext` via a small `learnServiceFromContext` helper, matching the pattern in `infrastructure/oss/unified_converter.go:194`. `DepLearnService` was already defined in `internal/context/context.go:102` so no new key was needed.

End-to-end: `withContext` → `deps[DepLearnService]=learnService` → `Scanner.Scan(ctx, …)` → `converter.ToIssues(ctx, …)` → `learnServiceFromContext(ctx)` → per-issue lookup. Code.Scanner SAST still uses the older field pattern; bringing it in line is a separate concern per `general.mdc` `"stick to the given goal"`.

File: infrastructure/secrets/secrets.go (hunk 3 above)
Comment (SnykOleg) [reply infrastructure/secrets/secrets.go]: Addressed in `3b238bb6` — the standalone `enhanceWithLearnLesson` loop is removed; lesson lookup now happens inside `FindingsConverter.findingToIssues` so each issue is enriched in the same pass that constructs it. No second walk over the issue slice. `ToIssues` takes a `context.Context` and reads the learn service once at the top, then a small `enrichWithLearnLesson(issue, ls)` helper does the per-issue mutation in-place. The nil-service / lookup-error / nil-lesson / empty-lesson branches all leave `LessonUrl` untouched so a Learn cache miss never fails the scan.

**Comment by rrama** [comment]
For the visuals, they are not docs related, right? They are just to show what it looks like for the PR? If so, we can just paste them in the PR description and let GitHub host them.

The test-release pipeline is failing, this seems to be because it doesn't handle manual non-semver tags (`edu-4754-pr1265-visuals-draft`), should be fine on merge, so we can ignore it.

**Comment by bastiandoetsch** [comment]
HI @SnykOleg , looks like not all commits are signed. Would you please sign all commits (or squash & sign) and re-push?

File: infrastructure/learn/service.go [hunk 4]
```diff
@@ -331,6 +331,15 @@ func (s *serviceImpl) lessonsLookupParams(
 			cwes,
 			cves,
 		}
+	case types.SecretsIssue:
```
Comment (basti-snyk) [inline infrastructure/learn/service.go:334]: **Should fix — a Secrets finding with no CWE gets an arbitrary, unrelated lesson.** This case passes an empty rule and relies on "CWE intersection through the all-lessons fall-through." That works only when the finding actually has a CWE. When it doesn't (`extractProblems` returns `cwes=nil` whenever there's no `cwe` problem), the lookup goes: empty rule + empty ecosystem → all-lessons branch → `filterForCWEs` with an empty CWE list, which `filterLessonWithComparatorFunc` treats as a no-op and returns the full list → `GetLesson` picks `lessons[0]`. So the "Learn how to remediate Secrets securely" link points at an arbitrary, unrelated lesson (and the selection is non-deterministic across runs, since `GetAll()` iterates a map).

Two reviewers flagged this independently. Suggested fix: in the Secrets path (or in `enrichWithLearnLesson`) skip the lookup / return no lesson when there is no CWE match, so a CWE-less secret gets *no* link rather than a wrong one. Add a real-service (non-mock) test asserting `LessonUrl` is empty for a no-CWE secret with a non-empty Learn cache — the current tests mock `GetLesson` and never exercise this path.

— AI review

File: infrastructure/secrets/convert.go [hunk 5]
```diff
@@ -128,11 +157,35 @@ func (c *FindingsConverter) findingToIssues(finding *testapi.FindingData, scanPa
 			FindingId:        attrs.Key,
 			Fingerprint:      attrs.Key,
 			AdditionalData:   additionalData,
-		})
+		}
+		c.enrichWithLearnLesson(issue, learnService)
```
Comment (basti-snyk) [inline infrastructure/secrets/convert.go:161]: **Suggestion — redundant per-location Learn lookups.** `enrichWithLearnLesson` is called inside the per-location loop, so a finding with N locations in a file emits N issues and issues N identical `GetLesson` calls (same ID/CWEs/CVEs/ecosystem). Since the Secrets path always misses the rule/ecosystem caches and falls through to a full lesson-catalog scan, each of those N calls scans the whole catalog. Resolve the lesson once per finding before the location loop and reuse the URL for every emitted issue.

Minor, non-blocking, related: `learnServiceFromContext` here duplicates the same `DepLearnService` context-read used in `oss/unified_converter.go` / `oss/cli_scanner.go` — worth a shared helper so the three copies can't drift.

— AI review

File: infrastructure/secrets/convert.go [hunk 6]
```diff
@@ -205,6 +258,22 @@ func (c *FindingsConverter) formattedMessageMarkdown(severity types.Severity, ti
 	return builder.String()
 }
 
+// countSourceLocationsInFile returns how many finding locations share the same
+// relative file path, for accurate "locations in this file" UI on per-location issues.
+func countSourceLocationsInFile(locations []testapi.FindingLocation, filePath string) int {
```
Comment (basti-snyk) [inline infrastructure/secrets/convert.go:263]: **Suggestion — `countSourceLocationsInFile` is O(L²) per finding.** It runs once per emitted issue (one per location), and each call loops all `L` locations and re-parses each via `AsSourceLocation()` (`json.Unmarshal`) — locations the outer `findingToIssues` loop already parsed once. On a finding with many same-file locations this is quadratic + redundant unmarshalling for a value that only depends on the finding. Consider a single O(L) pre-pass building `map[filePath]int` (reusing the parses the loop already does) before emitting issues. Non-blocking.

— AI review

File: infrastructure/secrets/secrets.go [hunk 7]
```diff
@@ -94,6 +97,23 @@ func New(conf configuration.Configuration, engine workflow.Engine, logger *zerol
 	}
 }
 
+// enrichContext injects the scanner-held learn service into the request context so
+// FindingsConverter can resolve LessonUrl on every scan path (including autosave and
+// background scans that use context.Background()). Mirrors infrastructure/oss/cli_scanner.go.
+func (sc *Scanner) enrichContext(ctx context.Context) context.Context {
```
Comment (basti-snyk) [inline infrastructure/secrets/secrets.go:103]: **Suggestion — the learn service is round-tripped through the request context for a single in-package caller.** `enrichContext` writes `sc.learnService` into the ctx deps map, and `ToIssues` reads it back out via `learnServiceFromContext`, yet the only caller of `ToIssues` (`Scan`, same file) already holds `sc.learnService` in scope. Passing it directly to the converter would drop `enrichContext`, `learnServiceFromContext`, the extra `context`/`learn`/`ctx2` imports, and a per-scan map copy. The Code path enriches as a post-scan pass over the deduped issue list (`infrastructure/code/code.go` `enhanceIssuesDetails`) rather than inside the per-location converter loop — following that shape here would also avoid the redundant per-location Learn lookups already noted. Non-blocking; weigh against any deliberate OSS-scanner parity.

— AI review
